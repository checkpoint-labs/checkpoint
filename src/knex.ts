import fs from 'fs';
import knex, { Knex } from 'knex';
import { ConnectionString } from 'connection-string';

export type KnexType =
  | {
      name: 'integer' | 'bigint' | 'boolean' | 'text' | 'json';
    }
  | {
      name: 'decimal';
      options: [number, number];
    }
  | {
      name: 'float';
      options: [number];
    }
  | {
      name: 'string';
      options: [number];
    };

const PROTOCOLS = {
  mysql: 'mysql',
  postgres: 'pg',
  postgresql: 'pg'
};

const EXTRA_OPTIONS = {
  mysql: {
    supportBigNumbers: true,
    bigNumberStrings: true
  }
};

export function getConnectionData(connectionString: string) {
  const connectionConfig = new ConnectionString(connectionString);
  if (!connectionConfig.protocol || !connectionConfig.hosts || !connectionConfig.path) {
    throw new Error('invalid connection string provided');
  }

  const client = PROTOCOLS[connectionConfig.protocol];
  if (!client) {
    throw new Error(`Supplied protocol ${connectionConfig.protocol} is not supported`);
  }

  const sslConfig: { rejectUnauthorized?: boolean; sslmode?: string; ca?: string } = {};
  if (
    connectionConfig.params?.sslaccept === 'strict' ||
    connectionConfig.params?.ssl === 'rejectUnauthorized'
  ) {
    sslConfig.rejectUnauthorized = true;
  }
  if (connectionConfig.params?.sslmode) {
    sslConfig.sslmode = connectionConfig.params.sslmode;
  }

  if (process.env.CA_CERT) {
    sslConfig.ca = process.env.CA_CERT;
  } else if (process.env.CA_CERT_FILE) {
    sslConfig.ca = fs.readFileSync(process.env.CA_CERT_FILE).toString();
  }

  return {
    client,
    connection: {
      database: connectionConfig.path[0],
      user: connectionConfig.user,
      password: connectionConfig.password,
      host: connectionConfig.hosts[0].name,
      port: connectionConfig.hosts[0].port,
      ssl: Object.keys(sslConfig).length > 0 ? sslConfig : undefined,
      ...EXTRA_OPTIONS[client]
    }
  };
}

export function createKnexConfig(connectionString: string): Knex.Config {
  return getConnectionData(connectionString);
}

export function createKnex(config: string | Knex.Config) {
  const parsedConfig = typeof config === 'string' ? createKnexConfig(config) : config;

  return knex(parsedConfig);
}
