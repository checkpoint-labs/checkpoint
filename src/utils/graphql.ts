import {
  GraphQLObjectType,
  GraphQLOutputType,
  GraphQLNonNull,
  isLeafType,
  isListType,
  GraphQLScalarType
} from 'graphql';
import { jsonToGraphQLQuery } from 'json-to-graphql-query';
import pluralize from 'pluralize';

export const extendSchema = (schema: string): string => {
  return `directive @derivedFrom(field: String!) on FIELD_DEFINITION
${schema}`;
};

/**
 * Returns name of query for fetching single entity record
 *
 */
export const singleEntityQueryName = (entity: GraphQLObjectType) => entity.name.toLowerCase();

/**
 * Returns name of query for fetching multiple entity records
 *
 */
export const multiEntityQueryName = (entity: GraphQLObjectType) => {
  if (entity.name === '_Metadata') return '_metadatas';

  return pluralize(entity.name.toLowerCase());
};

/**
 * Generate sample query string based on entity object fields.
 *
 */
export const generateQueryForEntity = (entity: GraphQLObjectType): string => {
  // function to recursively build fields map
  const getObjectFields = (object: GraphQLObjectType, queryFields = {}): Record<string, any> => {
    const objectFields = object.getFields();

    Object.keys(objectFields).forEach(fieldName => {
      const rawFieldType = objectFields[fieldName].type;
      const fieldType = rawFieldType instanceof GraphQLNonNull ? rawFieldType.ofType : rawFieldType;

      if (isLeafType(fieldType)) {
        queryFields[fieldName] = true;
      } else if (isListType(fieldType)) {
        if (fieldType.ofType instanceof GraphQLScalarType) {
          queryFields[fieldName] = true;
        }
      } else {
        const childObjectFields = {};
        getObjectFields(fieldType as GraphQLObjectType, childObjectFields);
        queryFields[fieldName] = childObjectFields;
      }
    });

    return queryFields;
  };

  return jsonToGraphQLQuery(
    {
      query: {
        [multiEntityQueryName(entity)]: {
          __args: { first: 10 },
          ...getObjectFields(entity)
        }
      }
    },
    { pretty: true }
  );
};

export const getNonNullType = (type: GraphQLOutputType): GraphQLOutputType => {
  if (type instanceof GraphQLNonNull) {
    return type.ofType;
  }

  return type;
};
