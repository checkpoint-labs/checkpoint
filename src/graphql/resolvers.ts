import { AsyncMySqlPool } from '../mysql';
import { Logger } from '../utils/logger';

/**
 *
 */
export interface ResolverContext {
  log: Logger;
  mysql: AsyncMySqlPool;
}

export async function queryMulti(parent, args, context: ResolverContext, info) {
  const { log, mysql } = context;

  const params: any = [];
  let whereSql = '';
  if (args.where) {
    Object.entries(args.where).map(w => {
      whereSql += !whereSql ? `WHERE ${w[0]} = ? ` : ` AND ${w[0]} = ?`;
      params.push(w[1]);
    });
  }
  const first = args?.first || 1000;
  const skip = args?.skip || 0;

  let orderBySql = '';
  if (args.orderBy) {
    orderBySql = `ORDER BY ${args.orderBy} ${args.orderDirection || 'DESC'}`;
  }

  params.push(skip, first);

  const query = `SELECT * FROM ${info.fieldName} ${whereSql} ${orderBySql} LIMIT ?, ?`;
  log.debug({ sql: query, args }, 'executing multi query');

  return await mysql.queryAsync(query, params);
}

export async function querySingle(parent, args, context: ResolverContext, info) {
  const { log, mysql } = context;

  const query = `SELECT * FROM ${info.fieldName}s WHERE id = ? LIMIT 1`;
  log.debug({ sql: query, args }, 'executing single query');

  const [item] = await mysql.queryAsync(query, [args.id]);
  return item;
}
