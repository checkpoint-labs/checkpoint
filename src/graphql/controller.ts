import {
  buildSchema,
  GraphQLEnumType,
  GraphQLField,
  GraphQLFieldConfig,
  GraphQLFieldConfigMap,
  GraphQLFieldResolver,
  GraphQLFloat,
  GraphQLID,
  GraphQLInputObjectType,
  GraphQLInputObjectTypeConfig,
  GraphQLInt,
  GraphQLList,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLOutputType,
  GraphQLScalarType,
  GraphQLSchema,
  GraphQLString,
  Source
} from 'graphql';
import { AsyncMySqlPool } from '../mysql';

import { querySingle, queryMulti, ResolverContext } from './resolvers';

/**
 * Type for single and multiple query resolvers
 */
interface EntityQueryResolvers<Context = ResolverContext> {
  singleEntityResolver: GraphQLFieldResolver<unknown, Context>;
  multipleEntityResolver: GraphQLFieldResolver<unknown, Context>;
}

const GraphQLOrderDirection = new GraphQLEnumType({
  name: 'OrderDirection',
  values: {
    asc: { value: 'ASC' },
    desc: { value: 'DESC' }
  }
});

/**
 * Controller for performing actions based on the graphql schema provided to its
 * constructor. It exposes public functions to generate graphql or database
 * items based on the entities identified in the schema.
 *
 * Note: Entities refer to Object types with an `id` field defined within the
 * graphql schema.
 */
export class GqlEntityController {
  private readonly schema: GraphQLSchema;

  constructor(typeDefs: string | Source) {
    this.schema = buildSchema(typeDefs);
  }

  /**
   * Creates a grqphql Query object generated from the objects defined within
   * the schema.
   * For each of the objects, two queries are created, one for querying the object
   * by it's id and the second for querying multiple objects based on a couple
   * of parameters.
   *
   * For example, given the input schema:
   * ```
   * type Vote {
   *  id: Int!
   *  name: String
   * }
   * ```
   *
   * The generated queries will be like:
   * ```
   * type Query {
   *  votes(
   *     first: Int
   *     skip: Int
   *     orderBy: String
   *     orderDirection: String
   *     where: WhereVote
   *   ): [Vote]
   *   vote(id: Int!): Vote
   * }
   *
   *  input WhereVote {
   *    id: Int
   *    id_in: [Int]
   *    name: String
   *    name_in: [String]
   *  }
   *
   * ```
   *
   */
  public generateQueryFields(
    schemaObjects?: GraphQLObjectType[],
    resolvers: EntityQueryResolvers = {
      singleEntityResolver: querySingle,
      multipleEntityResolver: queryMulti
    }
  ): GraphQLFieldConfigMap<any, any> {
    schemaObjects = schemaObjects || this.schemaObjects;

    const queryFields: GraphQLFieldConfigMap<any, any> = {};

    schemaObjects.forEach(type => {
      queryFields[type.name.toLowerCase()] = this.getSingleEntityQueryConfig(
        type,
        resolvers.singleEntityResolver
      );
      queryFields[`${type.name.toLowerCase()}s`] = this.getMultipleEntityQueryConfig(
        type,
        resolvers.multipleEntityResolver
      );
    });

    return queryFields;
  }

  /**
   * Creates store for each of the objects in the schema.
   * For now, it only creates mysql tables for each of the objects.
   * It also creates a checkpoint table to track checkpoints visited.
   *
   * For example, given an schema like:
   * ```graphql
   * type Vote {
   *  id: Int!
   *  name: String
   * }
   * ```
   *
   * will execute the following SQL:
   * ```sql
   * DROP TABLE IF EXISTS votes;
   * CREATE TABLE votes (
   *   id VARCHAR(128) NOT NULL,
   *   name VARCHAR(128),
   *   PRIMARY KEY (id) ,
   *   INDEX id (id),
   *   INDEX name (name)
   * );
   * ```
   *
   */
  public async createEntityStores(mysql: AsyncMySqlPool): Promise<void> {
    const schemaObjects = this.schemaObjects;
    if (schemaObjects.length === 0) {
      return;
    }

    let sql = '';
    schemaObjects.forEach(type => {
      sql += `\n\nDROP TABLE IF EXISTS ${type.name.toLowerCase()}s;`;
      sql += `\nCREATE TABLE ${type.name.toLowerCase()}s (`;
      let sqlIndexes = ``;

      this.getTypeFields(type).forEach(field => {
        const sqlType = this.getSqlType(field.type);

        sql += `\n  ${field.name} ${sqlType}`;
        if (field.type instanceof GraphQLNonNull) {
          sql += ' NOT NULL,';
        } else {
          sql += ',';
        }

        if (sqlType !== 'TEXT') {
          sqlIndexes += `,\n  INDEX ${field.name} (${field.name})`;
        }
      });
      sql += `\n  PRIMARY KEY (id) ${sqlIndexes}\n);\n`;
    });

    // TODO(perfectmak): wrap this in a transaction
    return mysql.queryAsync(sql.trimEnd());
  }

  public static generateQueryFields() {}

  /**
   * Returns a list of objects defined within the graphql typedefs.
   * The types returns are introspection objects, that can be used
   * for inspecting the fields and types.
   *
   * Note: that the returned objects does not include the Query object type if defined.
   *
   */
  private get schemaObjects(): GraphQLObjectType[] {
    return Object.values(this.schema.getTypeMap()).filter(type => {
      return (
        type instanceof GraphQLObjectType && type.name != 'Query' && !type.name.startsWith('__')
      );
    }) as GraphQLObjectType[];
  }

  private getTypeFields<Parent, Context>(
    type: GraphQLObjectType<Parent, Context>
  ): GraphQLField<Parent, Context>[] {
    return Object.values(type.getFields());
  }

  private getSingleEntityQueryConfig<Parent, Context>(
    type: GraphQLObjectType,
    resolver: GraphQLFieldResolver<Parent, Context>
  ): GraphQLFieldConfig<Parent, Context> {
    return {
      type,
      args: {
        id: { type: new GraphQLNonNull(this.getEntityIdType(type)) }
      },
      resolve: resolver
    };
  }

  private getMultipleEntityQueryConfig<Parent, Context>(
    type: GraphQLObjectType,
    resolver: GraphQLFieldResolver<Parent, Context>
  ): GraphQLFieldConfig<Parent, Context> {
    const whereInputConfig: GraphQLInputObjectTypeConfig = {
      name: `Where${type.name}`,
      fields: {}
    };

    this.getTypeFields(type).forEach(field => {
      // all field types in a where input variable must be optional
      // so we try to extract the non null type here.
      const nonNullFieldType = this.getNonNullType(field.type);

      // avoid setting up where filters for non scalar types
      if (!(nonNullFieldType instanceof GraphQLScalarType)) {
        return;
      }

      if (nonNullFieldType === GraphQLInt) {
        whereInputConfig.fields[`${field.name}_gt`] = { type: GraphQLInt };
        whereInputConfig.fields[`${field.name}_gte`] = { type: GraphQLInt };
        whereInputConfig.fields[`${field.name}_lt`] = { type: GraphQLInt };
        whereInputConfig.fields[`${field.name}_lte`] = { type: GraphQLInt };
      }

      if ((nonNullFieldType as GraphQLScalarType).name !== 'Text') {
        whereInputConfig.fields[`${field.name}`] = { type: nonNullFieldType };
        whereInputConfig.fields[`${field.name}_in`] = {
          type: new GraphQLList(nonNullFieldType)
        };
      }
    });

    return {
      type: new GraphQLList(type),
      args: {
        first: {
          type: GraphQLInt
        },
        skip: {
          type: GraphQLInt
        },
        orderBy: {
          type: GraphQLString
        },
        orderDirection: {
          type: GraphQLOrderDirection
        },
        where: { type: new GraphQLInputObjectType(whereInputConfig) }
      },
      resolve: resolver
    };
  }

  private getEntityIdType(type: GraphQLObjectType): GraphQLScalarType {
    const idField = type.getFields().id;
    if (!idField) {
      throw new Error(
        `'id' field is missing in type '${type.name}'. All types are required to have an id field.`
      );
    }

    if (!(idField.type instanceof GraphQLNonNull)) {
      throw new Error(`'id' field for type ${type.name} must be non nullable.`);
    }

    const nonNullType = idField.type.ofType;

    // verify only scalar types are used
    if (!(nonNullType instanceof GraphQLScalarType)) {
      throw new Error(`'id' field for type ${type.name} is not a scalar type.`);
    }

    return nonNullType;
  }

  private getNonNullType(type: GraphQLOutputType): GraphQLOutputType {
    if (type instanceof GraphQLNonNull) {
      return type.ofType;
    }

    return type;
  }

  /**
   * Return a mysql column type for the graphql type.
   *
   * It throws if the type is not a recognized scalar type.
   */
  private getSqlType(type: GraphQLOutputType): string {
    if (type instanceof GraphQLNonNull) {
      type = type.ofType;
    }

    switch (type) {
      case GraphQLInt:
        return 'INT(128)';
      case GraphQLFloat:
        return 'FLOAT(23)';
      case GraphQLString:
      case GraphQLID:
        return 'VARCHAR(128)';
    }

    // check for TEXT scalar type
    if (type instanceof GraphQLScalarType && type.name === 'Text') {
      return 'TEXT';
    }

    // TODO(perfectmak): Add support for List types

    throw new Error(`sql type for ${type} not support`);
  }
}
