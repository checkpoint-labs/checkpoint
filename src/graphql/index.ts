import { graphqlHTTP } from 'express-graphql';
import {
  GraphQLID,
  GraphQLInt,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString
} from 'graphql';
import { ResolverContext } from './resolvers';

/**
 * Creates an graphql http handler for the query passed a parameters.
 * Returned middleware can be used with express.
 */
export default function get(query: GraphQLObjectType, context: ResolverContext) {
  const schema = new GraphQLSchema({ query });
  return graphqlHTTP({ schema, context, graphiql: {} });
}

/**
 * This objects name and field maps to the values of the _metadata
 * database store
 *
 */
export const MetadataGraphQLObject = new GraphQLObjectType({
  name: '_Metadata',
  description: 'Core metadata values used internally by Checkpoint',
  fields: {
    id: { type: new GraphQLNonNull(GraphQLID), description: 'example: last_indexed_block' },
    value: { type: GraphQLString }
  }
});

/**
 * This objects name and field maps to the values of the _checkpoints
 * database store. And is used to generate entity queries for graphql
 *
 */
export const CheckpointsGraphQLObject = new GraphQLObjectType({
  name: '_Checkpoint',
  description: 'Contract and Block where its event is found.',
  fields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'id computed as last 5 bytes of sha256(contract+block)'
    },
    block_number: {
      type: new GraphQLNonNull(GraphQLInt)
    },
    contract_address: {
      type: new GraphQLNonNull(GraphQLString)
    }
  }
});
