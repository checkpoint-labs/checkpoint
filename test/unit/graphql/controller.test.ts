import { GraphQLObjectType, GraphQLSchema, printSchema } from 'graphql';
import knex from 'knex';
import { GqlEntityController } from '../../../src/graphql/controller';

describe('GqlEntityController', () => {
  describe('generateQueryFields', () => {
    it('should work', () => {
      const controller = new GqlEntityController(`
type Vote {
  id: Int!
  name: String
  authenticators: [String]
}
  `);
      const queryFields = controller.generateQueryFields();
      const querySchema = new GraphQLObjectType({
        name: 'Query',
        fields: queryFields
      });

      const schema = printSchema(new GraphQLSchema({ query: querySchema }));
      expect(schema).toMatchSnapshot();
    });

    // list of error table tests
    describe.each([
      {
        reason: 'non null object id',
        schema: `type Vote { id: String }`
      },
      {
        reason: 'object id is not scalar type',
        schema: `type Vote { id: Participant! }\n\n type Participant { id: Int! }`
      },
      {
        reason: 'object id is not scalar type 2',
        schema: `type Participant { id: [Int]! }`
      }
    ])('should fail for $reason', ({ schema }) => {
      const controller = new GqlEntityController(schema);
      expect(() => controller.generateQueryFields()).toThrowErrorMatchingSnapshot();
    });
  });

  describe('createEntityStores', () => {
    const mockKnex = knex({
      client: 'sqlite3',
      connection: {
        filename: ':memory:'
      },
      useNullAsDefault: true
    });

    afterAll(async () => {
      await mockKnex.destroy();
    });

    it('should work', async () => {
      const controller = new GqlEntityController(`
scalar BigInt
scalar Decimal
scalar BigDecimal

type Vote {
  id: Int!
  name: String
  authenticators: [String]
  big_number: BigInt
  decimal: Decimal
  big_decimal: BigDecimal
}
  `);
      const { builder } = await controller.createEntityStores(mockKnex);

      expect(builder.toString()).toMatchSnapshot();
    });
  });

  describe('generateSampleQuery', () => {
    it('should return undefined when no entities are defined', () => {
      const controller = new GqlEntityController('scalar Text');

      expect(controller.generateSampleQuery()).toBeUndefined();
    });

    it.each([
      {
        case: 'first and only entity',
        schema: `
type Vote {
  id: Int!
  name: String
  created_at: Int!
}
        `
      },
      {
        case: 'nested objects',
        // Checkpoint doesn't support relationship among entities,
        // but this just tests to ensure the sample query works for it.
        schema: `
type Vote {
  id: Int!
  name: String
  poster: Poster
  created_at: Int!
}

type Poster {
  id: Int!
  name: String!
  venue: Venue!
}

type Venue {
  id: Int!
  location: String!
}
        `
      }
    ])('should return correct query sample for $case', ({ schema }) => {
      const controller = new GqlEntityController(schema);

      expect(controller.generateSampleQuery()).not.toBeUndefined();
      expect(controller.generateSampleQuery()).toMatchSnapshot();
    });
  });
});
