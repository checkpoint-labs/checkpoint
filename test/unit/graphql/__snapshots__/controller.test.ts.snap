// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `"'id' field for type Vote must be non nullable."`;

exports[` 2`] = `"'id' field for type Vote is not a scalar type."`;

exports[` 3`] = `"'id' field for type Participant is not a scalar type."`;

exports[`GqlEntityController createEntityStores should work 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "

DROP TABLE IF EXISTS votes;
CREATE TABLE votes (
  id INT(128) NOT NULL,
  name VARCHAR(128),
  PRIMARY KEY (id) ,
  INDEX id (id),
  INDEX name (name)
);",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`GqlEntityController generateQueryFields should work 1`] = `
"type Query {
  vote(id: Int!): Vote
  votes(first: Int, skip: Int, orderBy: String, orderDirection: OrderDirection, where: WhereVote): [Vote]
}

type Vote {
  id: Int!
  name: String
}

enum OrderDirection {
  asc
  desc
}

input WhereVote {
  id_gt: Int
  id_gte: Int
  id_lt: Int
  id_lte: Int
  id: Int
  id_in: [Int]
  name: String
  name_in: [String]
}"
`;
